registry/remote/repository.go

```diff
@@ -944,61 +952,132 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
 }
 
 // generateDescriptor returns a descriptor generated from the response.
-func (s *manifestStore) generateDescriptor(resp *http.Response, ref registry.Reference) (ocispec.Descriptor, error) {
+// See the truth table at the top of `repository_test.go`
+func (s *manifestStore) generateDescriptor(resp *http.Response, ref registry.Reference, httpMethod string) (ocispec.Descriptor, error) {
+	// 1. Validate Content-Type
 	mediaType, _, err := mime.ParseMediaType(resp.Header.Get("Content-Type"))
 	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Content-Type: %w", resp.Request.Method, resp.Request.URL, err)
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: invalid response `Content-Type` header; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			err,
+		)
+	}
+
+	// 2. Validate Size
+	if resp.ContentLength == -1 {
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: unknown response Content-Length",
+			resp.Request.Method,
+			resp.Request.URL,
+		)
+	}
+
+	// 3. Validate Client Reference
+	var refDigest digest.Digest
+	if d, err := ref.Digest(); err == nil {
+		refDigest = d
+	}
+
+	// 4. Validate Server Digest (if present)
+	var serverHeaderDigest digest.Digest
+	if serverHeaderDigestStr := resp.Header.Get(dockerContentDigestHeader); serverHeaderDigestStr != "" {
+		if serverHeaderDigest, err = digest.Parse(serverHeaderDigestStr); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf(
+				"%s %q: invalid response header value `%s`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				dockerContentDigestHeader,
+				serverHeaderDigestStr,
+				err,
+			)
+		}
 	}
 
-	size := resp.ContentLength
-	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
-	}
+	/* 5. Now, look for specific error conditions; see truth table in method docstring */
+	var contentDigest digest.Digest
 
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+	var calculatedDigest digest.Digest
+	if len(serverHeaderDigest) == 0 {
+		if httpMethod == http.MethodHead {
+			if len(refDigest) == 0 {
+				// HEAD without server `Docker-Content-Digest` header is an
+				// immediate fail; HEAD
+				return ocispec.Descriptor{}, fmt.Errorf(
+					"HTTP %s request missing required header `%s`",
+					httpMethod, dockerContentDigestHeader,
+				)
+			}
+			// Otherwise, just trust the client-supplied digest
+			contentDigest = refDigest
+		} else {
+			// GET without server `Docker-Content-Digest` header forces the
+			// expensive calculation
+			if calculatedDigest, err = calculateDigestFromResponse(resp, s.repo.MaxMetadataBytes); err != nil {
+				return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+			}
```

Comment:
This suggestion reflects @Wwwsylvia 's [comment](https://github.com/oras-project/oras-go/pull/237/files#r954473588).

```suggestion
			calculatedDigest, err := calculateDigestFromResponse(resp, s.repo.MaxMetadataBytes)
			if err != nil {
				return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
			}
```

Reply:
with line 1001: `var calculatedDigest digest.Digest` removed

Reply:
Fixed.